src/System.Management.Automation/engine/PSVersionInfo.cs

```diff
@@ -571,77 +659,127 @@ public static bool TryParse(string version, out SemanticVersion result)
+                    versionSansLabel = version.Substring(0, dashIndex);
+                }
             }
 
-            var versionSansLabel = (dashIndex < 0) ? version : version.Substring(0, dashIndex);
-            string[] parsedComponents = versionSansLabel.Split(Utils.Separators.Dot);
-            if (parsedComponents.Length > 3)
+            if ((dashIndex != - 1 && String.IsNullOrEmpty(preLabel))   ||
+                (plusIndex != - 1 && String.IsNullOrEmpty(buildLabel)) ||
+                String.IsNullOrEmpty(versionSansLabel))
             {
-                result.SetFailure(ParseFailureKind.ArgumentException);
+                // We have dash and no preLabel or
+                // we have plus and no buildLabel or
+                // we have no main version part (versionSansLabel==null)
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            int major = 0, minor = 0, patch = 0;
-            if (!TryParseComponent(parsedComponents[0], "major", ref result, out major))
+            var match = Regex.Match(versionSansLabel, s_VersionSansRegEx);
+            if (!match.Success)
             {
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            if (parsedComponents.Length >= 2 && !TryParseComponent(parsedComponents[1], "minor", ref result, out minor))
+            if (!int.TryParse(match.Groups["major"].Value, out major))
             {
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            if (parsedComponents.Length == 3 && !TryParseComponent(parsedComponents[2], "patch", ref result, out patch))
+            if (match.Groups["minor"].Success && !int.TryParse(match.Groups["minor"].Value, out minor))
             {
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            result._parsedVersion = dashIndex < 0
-                ? new SemanticVersion(major, minor, patch)
-                : new SemanticVersion(major, minor, patch, version.Substring(dashIndex + 1));
-            return true;
-        }
-
-        private static bool TryParseComponent(string component, string componentName, ref VersionResult result, out int parsedComponent)
-        {
-            if (!Int32.TryParse(component, NumberStyles.Integer, CultureInfo.InvariantCulture, out parsedComponent))
+            if (match.Groups["patch"].Success && !int.TryParse(match.Groups["patch"].Value, out patch))
             {
-                result.SetFailure(ParseFailureKind.FormatException, component);
+                result.SetFailure(ParseFailureKind.FormatException);
                 return false;
             }
 
-            if (parsedComponent < 0)
+            if (preLabel != null && !Regex.IsMatch(preLabel, s_LabelUnitRegEx) ||
```

Comment:
I wonder if we should do this preLabel, buildLabel regex validation in the SemanticVersion constructor.  That way we can prevent malformed SemanticVersion objects from being created.

Reply:
I wonder too why I checked only first char in constructors :-) I wonted to be fast. We should use RegEx.

Fixed.